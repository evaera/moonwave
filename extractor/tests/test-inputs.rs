use std::path::Path;
use std::process::{Command, Stdio};

#[test]
fn everything_sandwich() -> anyhow::Result<()> {
    run_moonwave("passing/everything_sandwich.lua", 0)
}

#[test]
fn class_with_function() -> anyhow::Result<()> {
    run_moonwave("passing/class_with_function.lua", 0)
}

#[test]
fn fabric() -> anyhow::Result<()> {
    run_moonwave("passing/fabric.lua", 0)
}

#[test]
fn all_tags() -> anyhow::Result<()> {
    run_moonwave("passing/all_tags.lua", 0)
}

#[test]
fn indentation() -> anyhow::Result<()> {
    run_moonwave("passing/indentation.lua", 0)
}

#[test]
fn triple_dash() -> anyhow::Result<()> {
    run_moonwave("passing/triple_dash.lua", 0)
}

#[test]
fn free_function() -> anyhow::Result<()> {
    run_moonwave("passing/free_function.lua", 0)
}

#[test]
fn method_stmt_with_within() -> anyhow::Result<()> {
    run_moonwave("passing/method_stmt_with_within.lua", 0)
}

#[test]
fn class_with_index() -> anyhow::Result<()> {
    run_moonwave("passing/class_with_index.lua", 0)
}

#[test]
fn external_type() -> anyhow::Result<()> {
    run_moonwave("passing/external_types.lua", 0)
}

#[test]
fn drop_module() -> anyhow::Result<()> {
    run_moonwave("passing/drop_module.lua", 0)
}

#[test]
fn if_then_expression() -> anyhow::Result<()> {
    run_moonwave("passing/if_then_expression.lua", 0)
}
#[test]
fn docs_everywhere() -> anyhow::Result<()> {
    run_moonwave("passing/docs_everywhere.lua", 0)
}

#[test]
fn dynamic_param_return() -> anyhow::Result<()> {
    run_moonwave("passing/dynamic_param_return.lua", 0)
}

#[test]
fn triple_dash_wrong_comment() -> anyhow::Result<()> {
    run_moonwave("passing/triple_dash_wrong_comment.lua", 0)
}

#[test]
fn anomymous_function_assignment() -> anyhow::Result<()> {
    run_moonwave("passing/anonymous_function_assignment.lua", 0)
}

#[test]
fn type_statement_inference() -> anyhow::Result<()> {
    run_moonwave("passing/type_statement_inference.lua", 0)
}

#[test]
fn missing_fields() -> anyhow::Result<()> {
    run_moonwave("failing/missing_fields.lua", 1)
}

#[test]
fn failing_anomymous_function_assignment() -> anyhow::Result<()> {
    run_moonwave("failing/anonymous_function_assignment.lua", 1)
}

#[test]
fn param_validation() -> anyhow::Result<()> {
    run_moonwave("failing/param_validation.lua", 1)
}

#[test]
fn failing_function() -> anyhow::Result<()> {
    run_moonwave("failing/function.lua", 1)
}

#[test]
fn missing_starting_newline() -> anyhow::Result<()> {
    run_moonwave("failing/missing_starting_newline.lua", 1)
}

#[test]
fn failing_function_no_within() -> anyhow::Result<()> {
    run_moonwave("failing/function_no_within.lua", 1)
}

#[test]
fn class_with_unused_tags() -> anyhow::Result<()> {
    run_moonwave("failing/class_with_unused_tags.lua", 1)
}

#[test]
fn unknown_tags() -> anyhow::Result<()> {
    run_moonwave("failing/unknown_tags.lua", 1)
}

#[test]
fn duplicate_names() -> anyhow::Result<()> {
    run_moonwave("failing/duplicate_names.lua", 1)
}

fn run_moonwave(file_name: &str, expected_status: i32) -> anyhow::Result<()> {
    let path = Path::new("test-input").join(file_name);

    let child = Command::new(env!("CARGO_BIN_EXE_moonwave-extractor"))
        .arg("extract")
        .arg(path)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .env("NO_COLOR", "1")
        .spawn()?;

    let output = child.wait_with_output()?;
    let stdout = String::from_utf8(output.stdout)?;
    let stderr = String::from_utf8(output.stderr)?;

    let stdout_name = format!("{}-stdout", file_name);
    let stderr_name = format!("{}-stderr", file_name);

    let status_code = output.status.code();

    if status_code != Some(expected_status) {
        eprint!("{}", stderr);
        panic!(
            "Expected status code {}, but got {}",
            expected_status,
            status_code.map_or("none".to_owned(), |c| c.to_string())
        )
    }

    insta::assert_snapshot!(stdout_name, stdout);
    insta::assert_snapshot!(stderr_name, stderr);

    Ok(())
}
