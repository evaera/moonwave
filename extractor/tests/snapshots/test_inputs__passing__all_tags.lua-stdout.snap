---
source: tests/test-inputs.rs
expression: stdout
---
[
  {
    "functions": [
      {
        "name": "comingSoon",
        "desc": "",
        "params": [],
        "returns": [],
        "function_type": "static",
        "unreleased": true,
        "source": {
          "line": 57,
          "path": ""
        }
      }
    ],
    "properties": [
      {
        "name": "notReady",
        "desc": "",
        "lua_type": "boolean",
        "unreleased": true,
        "source": {
          "line": 77,
          "path": ""
        }
      }
    ],
    "types": [
      {
        "name": "nilOrVector",
        "desc": "We are still experimenting with vector technology.",
        "lua_type": "nil | vector",
        "unreleased": true,
        "source": {
          "line": 98,
          "path": ""
        }
      },
      {
        "name": "Command2",
        "desc": "We have fixed recursion! Unfortunately, it comes at the cost of groups.",
        "fields": [
          {
            "name": "Name",
            "lua_type": "string",
            "desc": "The name of the command."
          },
          {
            "name": "Recursion",
            "lua_type": "Command2",
            "desc": "No longer breaks the universe."
          },
          {
            "name": "Promise",
            "lua_type": "Promise",
            "desc": "A reference to the Promise library."
          },
          {
            "name": "Status",
            "lua_type": "Status",
            "desc": "The status of the internal promise."
          }
        ],
        "unreleased": true,
        "source": {
          "line": 132,
          "path": ""
        }
      }
    ],
    "name": "Bar",
    "desc": "",
    "unreleased": true,
    "source": {
      "line": 25,
      "path": ""
    }
  },
  {
    "functions": [
      {
        "name": "new",
        "desc": "This function creates a new Foo",
        "params": [
          {
            "name": "a",
            "desc": "param a",
            "lua_type": "A A A"
          },
          {
            "name": "b",
            "desc": "param b",
            "lua_type": "B B B"
          },
          {
            "name": "promise",
            "desc": "external param",
            "lua_type": "Promise"
          },
          {
            "name": "status",
            "desc": "external param",
            "lua_type": "Status"
          }
        ],
        "returns": [
          {
            "desc": "return a",
            "lua_type": "a"
          },
          {
            "desc": "return b",
            "lua_type": "b"
          },
          {
            "desc": "return external type",
            "lua_type": "Promise"
          },
          {
            "desc": "return external type",
            "lua_type": "Status"
          }
        ],
        "function_type": "static",
        "tags": [
          "uno",
          "dos"
        ],
        "external_types": [
          {
            "name": "Promise",
            "url": "https://eryn.io/roblox-lua-promise/api/Promise"
          },
          {
            "name": "Status",
            "url": "https://eryn.io/roblox-lua-promise/api/Promise#Status"
          }
        ],
        "errors": [
          {
            "lua_type": "c",
            "desc": "this errors sometimes. shrug"
          }
        ],
        "realm": [
          "Client",
          "Server"
        ],
        "since": "v1.23",
        "deprecated": {
          "version": "v1.5",
          "desc": "Use something else"
        },
        "private": true,
        "yields": true,
        "ignore": true,
        "source": {
          "line": 52,
          "path": ""
        }
      }
    ],
    "properties": [
      {
        "name": "ready",
        "desc": "",
        "lua_type": "boolean",
        "tags": [
          "salad",
          "fries",
          "nuggets"
        ],
        "realm": [
          "Client",
          "Server"
        ],
        "since": "v0.1.1",
        "deprecated": {
          "version": "v1.3",
          "desc": "Use blah"
        },
        "private": true,
        "readonly": true,
        "ignore": true,
        "source": {
          "line": 72,
          "path": ""
        }
      }
    ],
    "types": [
      {
        "name": "nilOrNumber",
        "desc": "Nil or number matey",
        "lua_type": "nil | number",
        "tags": [
          "arrgh",
          "yarr"
        ],
        "since": "v0.1.0",
        "deprecated": {
          "version": "v2.0.0",
          "desc": "We have finally removed all optional values from the library."
        },
        "private": true,
        "ignore": true,
        "source": {
          "line": 90,
          "path": ""
        }
      },
      {
        "name": "Command",
        "desc": "An object describing a command",
        "fields": [
          {
            "name": "Name",
            "lua_type": "string",
            "desc": "the name of the command"
          },
          {
            "name": "Groups",
            "lua_type": "array<string>",
            "desc": "A list of groups that the command contains"
          },
          {
            "name": "Recursion",
            "lua_type": "Command",
            "desc": "This breaks the universe"
          },
          {
            "name": "Promise",
            "lua_type": "Promise",
            "desc": "This is a Promise"
          },
          {
            "name": "Status",
            "lua_type": "Status",
            "desc": "Let's ignore that Promise:getStatus() exists"
          }
        ],
        "tags": [
          "cmdr",
          "lua"
        ],
        "external_types": [
          {
            "name": "OtherExample",
            "url": "www.example.com"
          },
          {
            "name": "AnotherExample",
            "url": "www.example.com"
          }
        ],
        "since": "v0.7.2",
        "deprecated": {
          "version": "v0.7.3",
          "desc": "Adding something that can break the universe was not such a good idea."
        },
        "private": true,
        "ignore": true,
        "source": {
          "line": 119,
          "path": ""
        }
      }
    ],
    "name": "Foo",
    "desc": "Here's a description for you\n\n:::info\nwith an admonition\n:::",
    "tags": [
      "bark",
      "meow",
      "rawr"
    ],
    "external_types": [
      {
        "name": "example",
        "url": "www.example.com"
      }
    ],
    "realm": [
      "Client",
      "Server",
      "Plugin"
    ],
    "deprecated": {
      "version": "v1.2",
      "desc": "Use x instead"
    },
    "since": "v1.0",
    "private": true,
    "ignore": true,
    "source": {
      "line": 21,
      "path": ""
    }
  }
]
